qutebrowser/config/configfiles.py

```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
         raise configexc.ConfigFileErrors(basename, [desc])
     except SyntaxError as e:
         desc = configexc.ConfigErrorDesc("Syntax Error", e,
-                                         traceback=traceback.format_exc())
+                                        traceback=traceback.format_exc())
         raise configexc.ConfigFileErrors(basename, [desc])
 
     try:
-        exec(code, module.__dict__)
+        # Save and restore sys variables
+        with saved_sys_properties():
+            # Add config directory to python path, so config.py can import
+            # other files in logical places
+            config_dir = os.path.dirname(filename)
+            if config_dir not in sys.path:
+                sys.path = [config_dir] + sys.path
+
+            exec(code, module.__dict__)
     except Exception as e:
         api.errors.append(configexc.ConfigErrorDesc(
             "Unhandled exception",
             exception=e, traceback=traceback.format_exc()))
-
     api.finalize()
     return api
 
 
+@contextlib.contextmanager
+def saved_sys_properties():
+    """Save various sys properties such as sys.path and sys.modules."""
+    old_path = sys.path
+    old_modules = sys.modules.copy()
+
+    try:
+        yield
+    except:
+        raise
+    finally:
+        sys.path = old_path
+        for module in set(sys.modules).difference(old_modules):
+            del sys.modules[module]
```

Comment:
I thought you could do:

```python
sys.modules.clear()
sys.modules.update(old_modules)
```

but Python doesn't like this:

```
>>> sys.modules.clear()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: lost builtins module
```

Reply:
Also, looking at my links again, I'm not sure anymore whether restoring `sys.modules` will actually solve or introduce more edge cases 😆 I *thought* I've seen this in some code I did read, but now I can't find it anymore... Did you see it somewhere?

Reply:
save/restoring of `sys.modules` is kind of a hack (since it dosen't properly clean up the modules). That blog post says that there's no better way to do this other than multiprocessing though, so I don't know if there's a better way (other than using `import` rather than `exec`).